# Embryo Stage Classification

This notebook trains four pretrained CNN backbones on the Kaggle embryo dataset:
- MobileNetV2
- VGG16
- VGG19
- InceptionV3

Each model is trained twice: once with weighted Cross Entropy (CE) and once with a custom Smooth Ordinal Loss that penalises stage-distance errors exponentially. Results from all eight experiments are compared at the end.

In [1]:
import torch

print(torch.cuda.is_available())
print(torch.version.cuda)
print(torch.cuda.get_device_name(0))

True
12.8
Tesla T4


In [2]:
import os
import copy
import random
from dataclasses import dataclass

import numpy as np
import pandas as pd
from tqdm.auto import tqdm

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader

from torchvision import transforms
from torchvision.models import (
    mobilenet_v2,
    vgg16,
    vgg19,
    inception_v3,
    MobileNet_V2_Weights,
    VGG16_Weights,
    VGG19_Weights,
    Inception_V3_Weights,
)

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score
from sklearn.utils.class_weight import compute_class_weight

from PIL import Image, ImageFile
ImageFile.LOAD_TRUNCATED_IMAGES = True

DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using device:', DEVICE)

Using device: cuda


## Configuration and Kaggle Dataset Paths

In [3]:
# -----------------------------------------------------------------------
# Kaggle dataset paths
# Add the dataset via:  Data -> Add Data -> search "embryo dataset"
# Dataset by abhishekbuddiga06 on Kaggle
# -----------------------------------------------------------------------
DATA_DIR = '/kaggle/input/datasets/abhishekbuddiga06/embryo-dataset/embryo_dataset/embryo_dataset'
ANN_DIR = '/kaggle/input/datasets/abhishekbuddiga06/embryo-dataset/embryo_dataset_annotations/embryo_dataset_annotations'
PHASES = [
    'tPB2', 'tPNa', 'tPNf', 't2', 't3', 't4', 't5', 't6',
    't7', 't8', 't9+', 'tM', 'tSB', 'tB', 'tEB', 'tHB'
]
label_map = {phase: idx for idx, phase in enumerate(PHASES)}

@dataclass
class Config:
    sample_rate: int = 8
    epochs: int = 10
    batch_size: int = 32
    inception_batch_size: int = 16
    image_size: int = 224
    inception_size: int = 299
    lr: float = 1e-4
    weight_decay: float = 1e-4
    num_workers: int = 2
    random_state: int = 42
    label_smoothing: float = 0.1
    grad_clip: float = 1.0
    # SmoothOrdinalLoss hyperparameters
    sol_alpha: float = 0.3
    sol_tau: float = 3.0
    aux_weight: float = 0.4
    merge_source: int = 15
    merge_target: int = 14

cfg = Config()
print(cfg)

Config(sample_rate=8, epochs=10, batch_size=32, inception_batch_size=16, image_size=224, inception_size=299, lr=0.0001, weight_decay=0.0001, num_workers=2, random_state=42, label_smoothing=0.1, grad_clip=1.0, sol_alpha=0.3, sol_tau=3.0, aux_weight=0.4, merge_source=15, merge_target=14)


In [4]:
def set_seed(seed: int) -> None:
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seed(cfg.random_state)

## Build Sampled Dataframe

Annotation CSV files map each embryo stage to a [start, end] frame range.
One frame is sampled every `sample_rate` frames within each annotated interval to reduce redundancy while preserving stage progression information.

In [5]:
import os

ANN_DIR = '/kaggle/input/datasets/abhishekbuddiga06/embryo-dataset/embryo_dataset_annotations/embryo_dataset_annotations'
print(os.listdir(ANN_DIR)[:5])

['BC750-7_phases.csv', 'SK308-7_phases.csv', 'QC697-4_phases.csv', 'DI358-3_phases.csv', 'DV210-4_phases.csv']


In [6]:
def build_dataframe(sample_rate: int = 8) -> pd.DataFrame:
    rows = []

    annotation_files = sorted([f for f in os.listdir(ANN_DIR) if f.endswith('.csv')])

    for file_name in tqdm(annotation_files, desc='Reading annotations'):
        embryo_id = file_name.replace('_phases.csv', '')
        ann_path  = os.path.join(ANN_DIR, file_name)
        image_dir = os.path.join(DATA_DIR, embryo_id)

        if not os.path.isdir(image_dir):
            continue

        image_files = sorted(os.listdir(image_dir))
        if not image_files:
            continue

        ann_df = pd.read_csv(ann_path, header=None, names=['phase', 'start', 'end'])

        for _, row in ann_df.iterrows():
            phase = row['phase']
            if phase not in label_map:
                continue

            start = int(row['start'])
            end   = int(row['end'])
            label = label_map[phase]

            for frame_idx in range(start, end, sample_rate):
                if 0 <= frame_idx < len(image_files):
                    rows.append({
                        'image':  os.path.join(image_dir, image_files[frame_idx]),
                        'label':  label,
                        'embryo': embryo_id,
                    })

    return pd.DataFrame(rows)

df = build_dataframe(cfg.sample_rate)
print('Total samples:', len(df))
display(df.head())

Reading annotations:   0%|          | 0/704 [00:00<?, ?it/s]

Total samples: 40337


,image,label,embryo
0,/kaggle/input/datasets/abhishekbuddiga06/embry...,0,AA83-7
1,/kaggle/input/datasets/abhishekbuddiga06/embry...,0,AA83-7
2,/kaggle/input/datasets/abhishekbuddiga06/embry...,0,AA83-7
3,/kaggle/input/datasets/abhishekbuddiga06/embry...,1,AA83-7
4,/kaggle/input/datasets/abhishekbuddiga06/embry...,1,AA83-7


In [7]:
# Split embryo-wise so no embryo appears in more than one split.
# This avoids data leakage through temporally correlated frames.
embryos = df['embryo'].unique()

train_ids, temp_ids = train_test_split(
    embryos, test_size=0.30, random_state=cfg.random_state, shuffle=True
)
val_ids, test_ids = train_test_split(
    temp_ids, test_size=0.50, random_state=cfg.random_state, shuffle=True
)

train_df = df[df['embryo'].isin(train_ids)].copy()
val_df   = df[df['embryo'].isin(val_ids)].copy()
test_df  = df[df['embryo'].isin(test_ids)].copy()

print('Split sizes:')
print('  Train:', len(train_df))
print('  Val  :', len(val_df))
print('  Test :', len(test_df))

Split sizes:
  Train: 28123
  Val  : 6156
  Test : 6058


In [8]:
# Class 15 (tHB) is merged into class 14 (tEB) because it has very few
# annotated samples and is visually the most similar to the preceding stage.
print('Train label counts before merge:')
display(train_df['label'].value_counts().sort_index())

for split_df in [train_df, val_df, test_df]:
    split_df.loc[split_df['label'] == cfg.merge_source, 'label'] = cfg.merge_target

num_classes = len(PHASES) - 1  # 15 effective classes

print('Train label counts after merge:')
display(train_df['label'].value_counts().sort_index())
print('Effective num_classes:', num_classes)

Train label counts before merge:


label
0      911
1     3907
2      770
3     2716
4      598
5     2714
6      862
7      898
8     1016
9     2990
10    4671
11    1620
12    1579
13     990
14    1877
15       4
Name: count, dtype: int64

Train label counts after merge:


label
0      911
1     3907
2      770
3     2716
4      598
5     2714
6      862
7      898
8     1016
9     2990
10    4671
11    1620
12    1579
13     990
14    1881
Name: count, dtype: int64

Effective num_classes: 15


In [9]:
# Compute inverse-frequency class weights for the CE baseline.
classes      = np.arange(num_classes)
weights      = compute_class_weight(class_weight='balanced', classes=classes, y=train_df['label'].values)
class_weights = torch.tensor(weights, dtype=torch.float32, device=DEVICE)

weight_table = pd.DataFrame({'class_id': classes, 'weight': weights})
display(weight_table)

,class_id,weight
0,0,2.058031
1,1,0.479874
2,2,2.434892
3,3,0.690304
4,4,3.135229
5,5,0.690813
6,6,2.175019
7,7,2.087825
8,8,1.845341
9,9,0.627046


## Dataset and Dataloaders

In [10]:
class EmbryoDataset(Dataset):
    def __init__(self, frame: pd.DataFrame, transform=None):
        self.frame     = frame.reset_index(drop=True)
        self.transform = transform

    def __len__(self):
        return len(self.frame)

    def __getitem__(self, idx):
        for _ in range(5):
            row = self.frame.iloc[idx]
            try:
                image = Image.open(row['image']).convert('RGB')
                if self.transform is not None:
                    image = self.transform(image)
                label = torch.tensor(int(row['label']), dtype=torch.long)
                return image, label
            except Exception:
                idx = random.randint(0, len(self.frame) - 1)

        fallback = torch.zeros(3, cfg.image_size, cfg.image_size)
        return fallback, torch.tensor(0, dtype=torch.long)


# Standard transforms (224x224) used by MobileNetV2, VGG16, VGG19
train_transform = transforms.Compose([
    transforms.Resize((cfg.image_size, cfg.image_size)),
    transforms.RandomHorizontalFlip(),
    transforms.RandomRotation(10),
    transforms.ColorJitter(brightness=0.1, contrast=0.1),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5]),
])

eval_transform = transforms.Compose([
    transforms.Resize((cfg.image_size, cfg.image_size)),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5]),
])

# InceptionV3 requires 299x299 input
inception_train_transform = transforms.Compose([
    transforms.Resize((cfg.inception_size, cfg.inception_size)),
    transforms.RandomHorizontalFlip(),
    transforms.RandomRotation(10),
    transforms.ColorJitter(brightness=0.1, contrast=0.1),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5]),
])

inception_eval_transform = transforms.Compose([
    transforms.Resize((cfg.inception_size, cfg.inception_size)),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5]),
])


def make_loader(frame, transform, batch_size, shuffle):
    return DataLoader(
        EmbryoDataset(frame, transform),
        batch_size=batch_size,
        shuffle=shuffle,
        num_workers=cfg.num_workers,
        pin_memory=torch.cuda.is_available(),
    )

standard_loaders = {
    'train': make_loader(train_df, train_transform,       cfg.batch_size,           True),
    'val':   make_loader(val_df,   eval_transform,        cfg.batch_size,           False),
    'test':  make_loader(test_df,  eval_transform,        cfg.batch_size,           False),
}

inception_loaders = {
    'train': make_loader(train_df, inception_train_transform, cfg.inception_batch_size, True),
    'val':   make_loader(val_df,   inception_eval_transform,  cfg.inception_batch_size, False),
    'test':  make_loader(test_df,  inception_eval_transform,  cfg.inception_batch_size, False),
}

print('Standard train batches :', len(standard_loaders['train']))
print('Inception train batches:', len(inception_loaders['train']))

Standard train batches : 879
Inception train batches: 1758


## Loss Functions

### Cross Entropy (Baseline)

Standard weighted cross entropy with label smoothing.  
Class weights derived from inverse class frequency to address the inherent class imbalance across embryo stages.

### Smooth Ordinal Loss (Custom)

Embryo development stages follow a strict biological order.  
Misclassifying stage t3 as t4 is a much smaller error than misclassifying it as tHB.  
Standard cross entropy treats both mistakes identically.

The custom loss adds an exponential distance penalty on top of CE:

$$\mathcal{L} = \mathcal{L}_{CE} + \alpha \cdot \frac{1}{N} \sum_{i=1}^{N} \left( \exp\left(\frac{|\hat{y}_{soft,i} - y_i|}{\tau}\right) - 1 \right)$$

where the soft prediction is:

$$\hat{y}_{soft} = \sum_{k=0}^{C-1} k \cdot p_k$$

- The soft prediction is differentiable, enabling direct gradient-based optimisation.  
- The exponential term means distant stage errors are penalised far more severely than close ones.  
- tau controls how quickly the penalty grows with distance.  
- alpha balances CE against the ordinal penalty.

In [11]:
class SmoothOrdinalLoss(nn.Module):
    """
    Cross Entropy plus an exponential ordinal-distance penalty.

    Adapted from the MAPE concept: rather than penalising percentage error,
    a soft expected-class index is computed from the predicted probability
    distribution and compared against the true class index.

    Parameters
    ----------
    alpha : float
        Weight of the distance penalty relative to CE.
    tau : float
        Temperature for the exponential penalty; larger values make the
        penalty grow more slowly with distance.
    """

    def __init__(self, alpha: float = 0.3, tau: float = 3.0):
        super().__init__()
        self.alpha = alpha
        self.tau   = tau
        self.ce    = nn.CrossEntropyLoss()

    def forward(self, outputs: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
        ce_loss = self.ce(outputs, targets)

        probs         = F.softmax(outputs, dim=1)
        class_indices = torch.arange(outputs.size(1), dtype=torch.float32, device=outputs.device)

        # Soft (differentiable) predicted class index
        y_pred_soft = torch.sum(probs * class_indices, dim=1)

        # Ordinal distance penalty
        distance = torch.abs(y_pred_soft - targets.float())
        penalty  = torch.exp(distance / self.tau) - 1.0

        return ce_loss + self.alpha * penalty.mean()


def make_loss(loss_name: str) -> nn.Module:
    if loss_name == 'ce':
        return nn.CrossEntropyLoss(
            weight=class_weights,
            label_smoothing=cfg.label_smoothing,
        ).to(DEVICE)

    if loss_name == 'smooth_ordinal':
        return SmoothOrdinalLoss(
            alpha=cfg.sol_alpha,
            tau=cfg.sol_tau,
        ).to(DEVICE)

    raise ValueError(f'Unknown loss: {loss_name}')

## Model Definitions

In [12]:
from torchvision.models import (
    mobilenet_v2, MobileNet_V2_Weights,
    vgg16, VGG16_Weights,
    vgg19, VGG19_Weights,
    inception_v3, Inception_V3_Weights
)
import torch.nn as nn

def get_model(name: str, num_classes: int) -> nn.Module:
    name = name.lower()

    if name == 'mobilenet':
        # ✅ load pretrained on CPU
        model = mobilenet_v2(weights=MobileNet_V2_Weights.DEFAULT)

        model.classifier = nn.Sequential(
            nn.Dropout(0.25),
            nn.Linear(model.last_channel, num_classes),
        )

    elif name == 'vgg16':
        model = vgg16(weights=VGG16_Weights.DEFAULT)

        model.classifier[6] = nn.Linear(
            model.classifier[6].in_features, num_classes
        )

    elif name == 'vgg19':
        model = vgg19(weights=VGG19_Weights.DEFAULT)

        model.classifier[6] = nn.Linear(
            model.classifier[6].in_features, num_classes
        )

    elif name == 'inception':
        model = inception_v3(
            weights=Inception_V3_Weights.DEFAULT,
            aux_logits=True
        )

        model.fc = nn.Linear(model.fc.in_features, num_classes)
        model.AuxLogits.fc = nn.Linear(
            model.AuxLogits.fc.in_features, num_classes
        )

    else:
        raise ValueError(f'Unknown model name: {name}')

    # ✅ move to GPU AFTER loading weights
    return model.to(DEVICE)

## Training and Evaluation Helpers

In [13]:
def unpack_outputs(outputs):
    """Handle both standard tensors and InceptionV3 named tuples."""
    if hasattr(outputs, 'logits'):
        return outputs.logits, getattr(outputs, 'aux_logits', None)
    if isinstance(outputs, tuple) and len(outputs) == 2:
        return outputs[0], outputs[1]
    return outputs, None


def train_one_epoch(model, loader, optimizer, loss_fn, model_name: str):
    model.train()
    running_loss = 0.0
    preds_all, targets_all = [], []

    progress = tqdm(loader, desc=f'{model_name} train', leave=False)
    for images, labels in progress:
        images = images.to(DEVICE, non_blocking=True)
        labels = labels.to(DEVICE, non_blocking=True)

        optimizer.zero_grad()
        outputs              = model(images)
        main_logits, aux_log = unpack_outputs(outputs)

        loss = loss_fn(main_logits, labels)
        if model_name == 'inception' and aux_log is not None:
            loss = loss + cfg.aux_weight * loss_fn(aux_log, labels)

        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), cfg.grad_clip)
        optimizer.step()

        running_loss += loss.item()
        preds = torch.argmax(main_logits, dim=1)
        preds_all.extend(preds.detach().cpu().numpy())
        targets_all.extend(labels.detach().cpu().numpy())
        progress.set_postfix(loss=f'{loss.item():.4f}')

    epoch_loss = running_loss / len(loader)
    epoch_acc  = accuracy_score(targets_all, preds_all)
    epoch_f1   = f1_score(targets_all, preds_all, average='weighted', zero_division=0)
    return epoch_loss, epoch_acc, epoch_f1


@torch.no_grad()
def evaluate(model, loader, model_name: str):
    model.eval()
    preds_all, targets_all = [], []

    for images, labels in tqdm(loader, desc=f'{model_name} eval', leave=False):
        images = images.to(DEVICE, non_blocking=True)
        labels = labels.to(DEVICE, non_blocking=True)

        outputs     = model(images)
        main_logits, _ = unpack_outputs(outputs)
        preds       = torch.argmax(main_logits, dim=1)

        preds_all.extend(preds.detach().cpu().numpy())
        targets_all.extend(labels.detach().cpu().numpy())

    acc = accuracy_score(targets_all, preds_all)
    f1  = f1_score(targets_all, preds_all, average='weighted', zero_division=0)
    return acc, f1

## Training Runner

In [14]:
def run_experiment(model_name: str, loss_name: str):
    print(f'\nTraining {model_name.upper()} | loss = {loss_name}')
    print('-' * 60)

    loaders = inception_loaders if model_name == 'inception' else standard_loaders
    model   = get_model(model_name, num_classes)

    optimizer = torch.optim.Adam(
        model.parameters(),
        lr=cfg.lr,
        weight_decay=cfg.weight_decay
    )

    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
        optimizer,
        mode='max',
        factor=0.5,
        patience=2
    )

    loss_fn = make_loss(loss_name)

    history     = []
    best_state  = None
    best_val_f1 = -1.0

    num_epochs = 10  # 🔥 increased epochs

    for epoch in range(1, num_epochs + 1):

        # ===== TRAIN =====
        train_loss, train_acc, train_f1 = train_one_epoch(
            model, loaders['train'], optimizer, loss_fn, model_name
        )

        # ===== VALIDATE =====
        val_acc, val_f1 = evaluate(model, loaders['val'], model_name)

        # ===== LR SCHEDULER =====
        old_lr = optimizer.param_groups[0]['lr']
        scheduler.step(val_f1)
        current_lr = optimizer.param_groups[0]['lr']

        if current_lr != old_lr:
            print(f'LR reduced: {old_lr:.2e} → {current_lr:.2e}')

        # ===== SAVE HISTORY =====
        history.append({
            'epoch':      epoch,
            'train_loss': train_loss,
            'train_acc':  train_acc,
            'train_f1':   train_f1,
            'val_acc':    val_acc,
            'val_f1':     val_f1,
            'lr':         current_lr,
        })

        # ===== PRINT =====
        print(
            f'Epoch {epoch:>2}/{num_epochs} | '
            f'Train Loss: {train_loss:.4f} | '
            f'Train Acc: {train_acc:.4f} | '
            f'Train F1: {train_f1:.4f} | '
            f'Val Acc: {val_acc:.4f} | '
            f'Val F1: {val_f1:.4f} | '
            f'LR: {current_lr:.2e}'
        )

        # ===== CHECKPOINT (BEST MODEL) =====
        if val_f1 > best_val_f1:
            best_val_f1 = val_f1
            best_state  = copy.deepcopy(model.state_dict())

            torch.save({
                'epoch': epoch,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'val_f1': val_f1,
            }, f'/kaggle/working/{model_name}_{loss_name}_best.pth')

    # ===== LOAD BEST MODEL =====
    if best_state is not None:
        model.load_state_dict(best_state)

    # ===== TEST =====
    test_acc, test_f1 = evaluate(model, loaders['test'], model_name)

    print(
        f'Best Val F1: {best_val_f1:.4f} | '
        f'Test Acc: {test_acc:.4f} | '
        f'Test F1: {test_f1:.4f}'
    )

    return pd.DataFrame(history), {
        'model':        model_name,
        'loss':         loss_name,
        'best_val_f1':  best_val_f1,
        'test_acc':     test_acc,
        'test_f1':      test_f1,
    }

## Experiment 1: Weighted Cross Entropy (Baseline)

All four models are trained using weighted cross entropy with label smoothing.

In [15]:
models_to_train = ['mobilenet', 'vgg16', 'vgg19', 'inception']

all_histories = []
all_results   = []

for model_name in models_to_train:
    hist_df, result = run_experiment(model_name, 'ce')
    hist_df['model'] = model_name
    hist_df['loss']  = 'ce'
    all_histories.append(hist_df)
    all_results.append(result)


Training MOBILENET | loss = ce
------------------------------------------------------------
Downloading: "https://download.pytorch.org/models/mobilenet_v2-7ebf99e0.pth" to /root/.cache/torch/hub/checkpoints/mobilenet_v2-7ebf99e0.pth


100%|██████████| 13.6M/13.6M [00:00<00:00, 116MB/s] 


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  1/10 | Train Loss: 2.3871 | Train Acc: 0.2313 | Train F1: 0.2425 | Val Acc: 0.2752 | Val F1: 0.2797 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.1848 | Train Acc: 0.3021 | Train F1: 0.3147 | Val Acc: 0.3018 | Val F1: 0.3143 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^Exception ignored in: ^^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>

  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    Traceback (most recent call last):
assert self._parent_pid == os.getpid(), 'can only test a child process'  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

     self._shutdown_workers() 
   File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
      if w.is_alive(): 
       ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^

Epoch  3/10 | Train Loss: 2.1054 | Train Acc: 0.3318 | Train F1: 0.3433 | Val Acc: 0.2786 | Val F1: 0.2870 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Exception ignored in: 
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Traceback (most recent call last):
^
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
Traceback (most recent call last):
      File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers

      File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():if w.is_alive():

            ^^^ ^ ^^^^^^^^^Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>^^^^
Traceback (most recent call last):
^
  File "/usr/local/lib/python3.12/dist-packages/torch/util

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 2.0314 | Train Acc: 0.3649 | Train F1: 0.3767 | Val Acc: 0.2786 | Val F1: 0.2825 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  5/10 | Train Loss: 1.9533 | Train Acc: 0.3958 | Train F1: 0.4063 | Val Acc: 0.2857 | Val F1: 0.2991 | LR: 5.00e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  6/10 | Train Loss: 1.8627 | Train Acc: 0.4406 | Train F1: 0.4509 | Val Acc: 0.2827 | Val F1: 0.2987 | LR: 5.00e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  7/10 | Train Loss: 1.8107 | Train Acc: 0.4600 | Train F1: 0.4702 | Val Acc: 0.2817 | Val F1: 0.2939 | LR: 5.00e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

LR reduced: 5.00e-05 → 2.50e-05
Epoch  8/10 | Train Loss: 1.7692 | Train Acc: 0.4847 | Train F1: 0.4946 | Val Acc: 0.2947 | Val F1: 0.3098 | LR: 2.50e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Exception ignored in: 
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Traceback (most recent call last):

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
        self._shutdown_workers()
self._shutdown_workers()  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
        if w.is_alive():if w.is_alive():

           ^ ^ ^ ^^^^^^^^^^^^^^^^^^
^  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'^^

   File "/usr/lib/pytho

Epoch  9/10 | Train Loss: 1.7079 | Train Acc: 0.5130 | Train F1: 0.5230 | Val Acc: 0.2859 | Val F1: 0.2980 | LR: 2.50e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
Exception ignored in:   File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
    Traceback (most recent call last):
if w.is_alive():  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

      self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
       if w.is_alive(): 
   ^^ ^^^ ^  ^ ^^^^^^Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>^^

Traceback (most recent call last):
^  File "/usr/lib/python3.12/multiprocessing/process.py", line 

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 1.6880 | Train Acc: 0.5226 | Train F1: 0.5323 | Val Acc: 0.2939 | Val F1: 0.3078 | LR: 2.50e-05


mobilenet eval:   0%|          | 0/190 [00:00<?, ?it/s]

Best Val F1: 0.3143 | Test Acc: 0.2956 | Test F1: 0.3081

Training VGG16 | loss = ce
------------------------------------------------------------
Downloading: "https://download.pytorch.org/models/vgg16-397923af.pth" to /root/.cache/torch/hub/checkpoints/vgg16-397923af.pth


100%|██████████| 528M/528M [00:02<00:00, 190MB/s]  


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  1/10 | Train Loss: 2.3987 | Train Acc: 0.2292 | Train F1: 0.2437 | Val Acc: 0.2468 | Val F1: 0.2488 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.2369 | Train Acc: 0.2743 | Train F1: 0.2893 | Val Acc: 0.2409 | Val F1: 0.2366 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 2.1789 | Train Acc: 0.2857 | Train F1: 0.2991 | Val Acc: 0.2926 | Val F1: 0.2965 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

Epoch  4/10 | Train Loss: 2.1326 | Train Acc: 0.3020 | Train F1: 0.3149 | Val Acc: 0.2584 | Val F1: 0.2696 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Exception ignored in: 
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
        self._shutdown_workers()self._shutdown_workers()

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
        if w.is_alive():if w.is_alive():
 
            ^ ^^^^^^^^^^^^^^^^^^^^^^
^  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive

  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
        assert self.

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  5/10 | Train Loss: 2.0808 | Train Acc: 0.3201 | Train F1: 0.3318 | Val Acc: 0.2477 | Val F1: 0.2591 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  6/10 | Train Loss: 2.0391 | Train Acc: 0.3403 | Train F1: 0.3526 | Val Acc: 0.2526 | Val F1: 0.2603 | LR: 5.00e-05


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  7/10 | Train Loss: 1.9318 | Train Acc: 0.3803 | Train F1: 0.3929 | Val Acc: 0.2617 | Val F1: 0.2795 | LR: 5.00e-05


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  8/10 | Train Loss: 1.8803 | Train Acc: 0.4077 | Train F1: 0.4191 | Val Acc: 0.2935 | Val F1: 0.3089 | LR: 5.00e-05


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  9/10 | Train Loss: 1.8347 | Train Acc: 0.4315 | Train F1: 0.4441 | Val Acc: 0.2823 | Val F1: 0.2953 | LR: 5.00e-05


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

Epoch 10/10 | Train Loss: 1.7963 | Train Acc: 0.4533 | Train F1: 0.4656 | Val Acc: 0.2989 | Val F1: 0.3155 | LR: 5.00e-05


vgg16 eval:   0%|          | 0/190 [00:00<?, ?it/s]

Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
Exception ignored in:   File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
    Traceback (most recent call last):
self._shutdown_workers()  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

      File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
self._shutdown_workers()
      File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
if w.is_alive():
     if w.is_alive():  
      ^ ^  ^ ^ ^^^^^^^^^^^^^^^^^
^  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
^^
      File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
assert self._pa

Best Val F1: 0.3155 | Test Acc: 0.2923 | Test F1: 0.3023

Training VGG19 | loss = ce
------------------------------------------------------------
Downloading: "https://download.pytorch.org/models/vgg19-dcbb9e9d.pth" to /root/.cache/torch/hub/checkpoints/vgg19-dcbb9e9d.pth


100%|██████████| 548M/548M [00:03<00:00, 168MB/s]  


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  1/10 | Train Loss: 2.3982 | Train Acc: 0.2233 | Train F1: 0.2365 | Val Acc: 0.2336 | Val F1: 0.2378 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.2460 | Train Acc: 0.2720 | Train F1: 0.2856 | Val Acc: 0.2081 | Val F1: 0.1822 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 2.1952 | Train Acc: 0.2791 | Train F1: 0.2915 | Val Acc: 0.2302 | Val F1: 0.2429 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 2.1521 | Train Acc: 0.2962 | Train F1: 0.3096 | Val Acc: 0.2355 | Val F1: 0.2446 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    Exception ignored in: self._shutdown_workers()<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers

    Traceback (most recent call last):
if w.is_alive():  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

     self._shutdown_workers()
    File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
      if w.is_alive(): 
  ^^ ^ ^  ^  ^^^^^^^^^^^^^^
^  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    ^^assert self._parent_pid == os.getpid(), 'can only test a child process'^
^  
   File "/usr/lib/pyt

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Exception ignored in: Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Traceback (most recent call last):
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

    Traceback (most recent call last):
Traceback (most recent call last):
self._shutdown_workers()  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

      File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    self._shutdown_workers()self._shutdown_workers()    

Exception ignored in: if w.is_alive():  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloade

Epoch  5/10 | Train Loss: 2.1230 | Train Acc: 0.3058 | Train F1: 0.3183 | Val Acc: 0.2885 | Val F1: 0.2972 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  6/10 | Train Loss: 2.0820 | Train Acc: 0.3182 | Train F1: 0.3296 | Val Acc: 0.2544 | Val F1: 0.2525 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  7/10 | Train Loss: 2.0559 | Train Acc: 0.3343 | Train F1: 0.3463 | Val Acc: 0.2675 | Val F1: 0.2752 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  8/10 | Train Loss: 2.0217 | Train Acc: 0.3458 | Train F1: 0.3568 | Val Acc: 0.2861 | Val F1: 0.2901 | LR: 5.00e-05


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  9/10 | Train Loss: 1.9165 | Train Acc: 0.3868 | Train F1: 0.3967 | Val Acc: 0.2749 | Val F1: 0.2843 | LR: 5.00e-05


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 1.8758 | Train Acc: 0.4074 | Train F1: 0.4186 | Val Acc: 0.2823 | Val F1: 0.3043 | LR: 5.00e-05


vgg19 eval:   0%|          | 0/190 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
        Exception ignored in:  <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0> 
 Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
^    self._shutdown_workers()^^
^^  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
^    if w.is_alive():^
^ ^^^^  ^^ ^^ ^  ^^^^^^^^

Best Val F1: 0.3043 | Test Acc: 0.2767 | Test F1: 0.2913

Training INCEPTION | loss = ce
------------------------------------------------------------
Downloading: "https://download.pytorch.org/models/inception_v3_google-0cc3c7bd.pth" to /root/.cache/torch/hub/checkpoints/inception_v3_google-0cc3c7bd.pth


100%|██████████| 104M/104M [00:00<00:00, 179MB/s] 


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  1/10 | Train Loss: 3.2818 | Train Acc: 0.2519 | Train F1: 0.2666 | Val Acc: 0.2661 | Val F1: 0.2519 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 3.0687 | Train Acc: 0.3092 | Train F1: 0.3227 | Val Acc: 0.2357 | Val F1: 0.2359 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

    Traceback (most recent call last):
self._shutdown_workers()  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
        self._shutdown_workers()if w.is_alive():

   File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
      if w.is_alive(): 
    ^ ^  ^ ^ ^^ ^^^^^^^^^^^
^^  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
^    assert self._parent_pid == os.getpid(), 'can only test a child process'^
 ^ ^ ^  
   File "/usr/lib/

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 2.9458 | Train Acc: 0.3399 | Train F1: 0.3535 | Val Acc: 0.2571 | Val F1: 0.2646 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 2.8314 | Train Acc: 0.3776 | Train F1: 0.3912 | Val Acc: 0.3345 | Val F1: 0.3397 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Exception ignored in: 
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Traceback (most recent call last):

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
        self._shutdown_workers()self._shutdown_workers()

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
        if w.is_alive():if w.is_alive():

              ^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._par

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  5/10 | Train Loss: 2.7468 | Train Acc: 0.4050 | Train F1: 0.4174 | Val Acc: 0.3109 | Val F1: 0.3190 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  6/10 | Train Loss: 2.6554 | Train Acc: 0.4350 | Train F1: 0.4472 | Val Acc: 0.2653 | Val F1: 0.2849 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  7/10 | Train Loss: 2.5877 | Train Acc: 0.4642 | Train F1: 0.4773 | Val Acc: 0.2716 | Val F1: 0.2847 | LR: 5.00e-05


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  8/10 | Train Loss: 2.3613 | Train Acc: 0.5433 | Train F1: 0.5549 | Val Acc: 0.3246 | Val F1: 0.3401 | LR: 5.00e-05


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  9/10 | Train Loss: 2.2624 | Train Acc: 0.5833 | Train F1: 0.5944 | Val Acc: 0.3337 | Val F1: 0.3510 | LR: 5.00e-05


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 2.1948 | Train Acc: 0.6101 | Train F1: 0.6200 | Val Acc: 0.3255 | Val F1: 0.3380 | LR: 5.00e-05


inception eval:   0%|          | 0/379 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>Exception ignored in: 
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Exception ignored in: Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__

    self._shutdown_workers()Traceback (most recent call last):
    
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
self._shutdown_workers()        
if w.is_alive():self._shutdown_workers()  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _s

Best Val F1: 0.3510 | Test Acc: 0.3320 | Test F1: 0.3446


## Experiment 2: Smooth Ordinal Loss (Custom)

All four models are retrained using the custom SmoothOrdinalLoss.

In [16]:
for model_name in models_to_train:
    hist_df, result = run_experiment(model_name, 'smooth_ordinal')
    hist_df['model'] = model_name
    hist_df['loss']  = 'smooth_ordinal'
    all_histories.append(hist_df)
    all_results.append(result)


Training MOBILENET | loss = smooth_ordinal
------------------------------------------------------------


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>^
Traceback (most recent call last):

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
      File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    self._shutdown_workers()
assert self._parent_pid == os.getpid(), 'can only test a child process'  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers

    if w.is_alive(): 
               ^ ^ ^^^^^^^^^^^^^^^^^^^
^  

Epoch  1/10 | Train Loss: 2.3879 | Train Acc: 0.3556 | Train F1: 0.2993 | Val Acc: 0.3653 | Val F1: 0.3222 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.0465 | Train Acc: 0.4080 | Train F1: 0.3530 | Val Acc: 0.3938 | Val F1: 0.3474 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 1.9098 | Train Acc: 0.4395 | Train F1: 0.3866 | Val Acc: 0.3939 | Val F1: 0.3424 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 1.7965 | Train Acc: 0.4633 | Train F1: 0.4159 | Val Acc: 0.3861 | Val F1: 0.3368 | LR: 1.00e-04


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  5/10 | Train Loss: 1.6868 | Train Acc: 0.4896 | Train F1: 0.4456 | Val Acc: 0.3798 | Val F1: 0.3367 | LR: 5.00e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  6/10 | Train Loss: 1.5320 | Train Acc: 0.5260 | Train F1: 0.4867 | Val Acc: 0.3822 | Val F1: 0.3429 | LR: 5.00e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>^^
^^Traceback (most recent call last):
^^  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
^^    self._shutdown_workers()^^
^  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
^    ^if w.is_alive():
^  ^ ^ ^ 

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  7/10 | Train Loss: 1.4533 | Train Acc: 0.5456 | Train F1: 0.5099 | Val Acc: 0.3785 | Val F1: 0.3408 | LR: 5.00e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>can only test a child process

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0><function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>

Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
        self._shutdown_workers()self._shutdown_workers()

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
        if w.is_alive():if w.is_alive():
 
        Exception ignored in:   <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>   
^^Exception ignored in: Traceback (most recent call last):
^^<function _MultiProcessingDataLoaderIter.__del__ at 0

LR reduced: 5.00e-05 → 2.50e-05
Epoch  8/10 | Train Loss: 1.4084 | Train Acc: 0.5616 | Train F1: 0.5305 | Val Acc: 0.3770 | Val F1: 0.3456 | LR: 2.50e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  9/10 | Train Loss: 1.3051 | Train Acc: 0.5885 | Train F1: 0.5598 | Val Acc: 0.3788 | Val F1: 0.3462 | LR: 2.50e-05


mobilenet train:   0%|          | 0/879 [00:00<?, ?it/s]

mobilenet eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 1.2666 | Train Acc: 0.5988 | Train F1: 0.5718 | Val Acc: 0.3806 | Val F1: 0.3476 | LR: 2.50e-05


mobilenet eval:   0%|          | 0/190 [00:00<?, ?it/s]

Best Val F1: 0.3476 | Test Acc: 0.3797 | Test F1: 0.3466

Training VGG16 | loss = smooth_ordinal
------------------------------------------------------------


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^Exception ignored in: ^^^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>

  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
Traceback (most recent call last):
      File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
assert self._parent_pid == os.getpid(), 'can only test a child process'    
self._shutdown_workers() 
   File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
       if w.is_alive():
         ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^^

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  1/10 | Train Loss: 2.3707 | Train Acc: 0.3480 | Train F1: 0.2961 | Val Acc: 0.3476 | Val F1: 0.2923 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.1004 | Train Acc: 0.3941 | Train F1: 0.3374 | Val Acc: 0.3787 | Val F1: 0.3200 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 2.0051 | Train Acc: 0.4105 | Train F1: 0.3527 | Val Acc: 0.3769 | Val F1: 0.3157 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 1.9088 | Train Acc: 0.4288 | Train F1: 0.3715 | Val Acc: 0.3905 | Val F1: 0.3250 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  5/10 | Train Loss: 1.8438 | Train Acc: 0.4441 | Train F1: 0.3879 | Val Acc: 0.4046 | Val F1: 0.3470 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

Epoch  6/10 | Train Loss: 1.7773 | Train Acc: 0.4586 | Train F1: 0.4067 | Val Acc: 0.3926 | Val F1: 0.3370 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>    
self._shutdown_workers()Traceback (most recent call last):

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
        if w.is_alive():self._shutdown_workers()
 
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
       if w.is_alive():
 Exception ignored in:  <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>   
^ ^ Traceback (most recent call last):
^  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  7/10 | Train Loss: 1.7141 | Train Acc: 0.4722 | Train F1: 0.4209 | Val Acc: 0.3761 | Val F1: 0.3214 | LR: 1.00e-04


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  8/10 | Train Loss: 1.6500 | Train Acc: 0.4872 | Train F1: 0.4369 | Val Acc: 0.3848 | Val F1: 0.3298 | LR: 5.00e-05


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  9/10 | Train Loss: 1.4735 | Train Acc: 0.5283 | Train F1: 0.4833 | Val Acc: 0.3960 | Val F1: 0.3559 | LR: 5.00e-05


vgg16 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg16 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 1.4005 | Train Acc: 0.5470 | Train F1: 0.5075 | Val Acc: 0.3968 | Val F1: 0.3618 | LR: 5.00e-05


vgg16 eval:   0%|          | 0/190 [00:00<?, ?it/s]

Best Val F1: 0.3618 | Test Acc: 0.3747 | Test F1: 0.3353

Training VGG19 | loss = smooth_ordinal
------------------------------------------------------------


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    Exception ignored in: self._shutdown_workers()
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    
if w.is_alive():Traceback (most recent call last):

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
       self._shutdown_workers() 
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
       ^if w.is_alive():^
^ ^ ^  ^ ^ ^ ^^^^^^^^
^^  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
^ ^ ^  ^^^ 
    File "/usr/lib

Epoch  1/10 | Train Loss: 2.3928 | Train Acc: 0.3458 | Train F1: 0.2938 | Val Acc: 0.3764 | Val F1: 0.3104 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.1318 | Train Acc: 0.3893 | Train F1: 0.3329 | Val Acc: 0.3884 | Val F1: 0.3215 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 2.0440 | Train Acc: 0.4058 | Train F1: 0.3489 | Val Acc: 0.3739 | Val F1: 0.3109 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 1.9732 | Train Acc: 0.4173 | Train F1: 0.3596 | Val Acc: 0.3725 | Val F1: 0.3049 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  5/10 | Train Loss: 1.9212 | Train Acc: 0.4260 | Train F1: 0.3675 | Val Acc: 0.3861 | Val F1: 0.3285 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  6/10 | Train Loss: 1.8732 | Train Acc: 0.4344 | Train F1: 0.3760 | Val Acc: 0.3748 | Val F1: 0.3202 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^Exception ignored in: ^<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>^
^Traceback (most recent call last):
^^^  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
^^    ^self._shutdown_workers()^
^^  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
^^    ^if w.is_alive():^
^^ ^ ^ ^ ^^^ 

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    Exception ignored in: self._shutdown_workers()<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>

  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
Traceback (most recent call last):
    if w.is_alive():  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
Exception ignored in: 
<function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>     self._shutdown_workers()

 Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
   File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
        

Epoch  7/10 | Train Loss: 1.8341 | Train Acc: 0.4452 | Train F1: 0.3861 | Val Acc: 0.3892 | Val F1: 0.3131 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  8/10 | Train Loss: 1.7852 | Train Acc: 0.4565 | Train F1: 0.3983 | Val Acc: 0.4022 | Val F1: 0.3413 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch  9/10 | Train Loss: 1.7336 | Train Acc: 0.4634 | Train F1: 0.4073 | Val Acc: 0.3957 | Val F1: 0.3450 | LR: 1.00e-04


vgg19 train:   0%|          | 0/879 [00:00<?, ?it/s]

vgg19 eval:   0%|          | 0/193 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 1.6962 | Train Acc: 0.4740 | Train F1: 0.4184 | Val Acc: 0.3980 | Val F1: 0.3451 | LR: 1.00e-04


vgg19 eval:   0%|          | 0/190 [00:00<?, ?it/s]

Best Val F1: 0.3451 | Test Acc: 0.3810 | Test F1: 0.3256

Training INCEPTION | loss = smooth_ordinal
------------------------------------------------------------


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  1/10 | Train Loss: 3.1784 | Train Acc: 0.3686 | Train F1: 0.3149 | Val Acc: 0.3829 | Val F1: 0.3273 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  2/10 | Train Loss: 2.8108 | Train Acc: 0.4189 | Train F1: 0.3660 | Val Acc: 0.3826 | Val F1: 0.3370 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  3/10 | Train Loss: 2.6035 | Train Acc: 0.4504 | Train F1: 0.4020 | Val Acc: 0.3941 | Val F1: 0.3445 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  4/10 | Train Loss: 2.4099 | Train Acc: 0.4879 | Train F1: 0.4460 | Val Acc: 0.3928 | Val F1: 0.3452 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  5/10 | Train Loss: 2.2443 | Train Acc: 0.5119 | Train F1: 0.4734 | Val Acc: 0.3785 | Val F1: 0.3289 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  6/10 | Train Loss: 2.1226 | Train Acc: 0.5370 | Train F1: 0.5047 | Val Acc: 0.3936 | Val F1: 0.3553 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  7/10 | Train Loss: 2.0377 | Train Acc: 0.5587 | Train F1: 0.5283 | Val Acc: 0.3928 | Val F1: 0.3531 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch  8/10 | Train Loss: 1.9414 | Train Acc: 0.5760 | Train F1: 0.5480 | Val Acc: 0.3692 | Val F1: 0.3415 | LR: 1.00e-04


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

LR reduced: 1.00e-04 → 5.00e-05
Epoch  9/10 | Train Loss: 1.8655 | Train Acc: 0.5890 | Train F1: 0.5645 | Val Acc: 0.3866 | Val F1: 0.3527 | LR: 5.00e-05


inception train:   0%|          | 0/1758 [00:00<?, ?it/s]

Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1690, in _shutdown_workers
    if w.is_alive():
       ^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/process.py", line 160, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: can only test a child process
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7af5455ae2a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 1707, in __del__
    self._shutdown_workers()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py", line 16

inception eval:   0%|          | 0/385 [00:00<?, ?it/s]

Epoch 10/10 | Train Loss: 1.5419 | Train Acc: 0.6544 | Train F1: 0.6342 | Val Acc: 0.3808 | Val F1: 0.3584 | LR: 5.00e-05


inception eval:   0%|          | 0/379 [00:00<?, ?it/s]

Best Val F1: 0.3584 | Test Acc: 0.3856 | Test F1: 0.3621


## Final Results

In [21]:
history_table = pd.concat(all_histories, ignore_index=True)
results_table = (
    pd.DataFrame(all_results)
    .sort_values(['loss', 'test_f1'], ascending=[True, False])
    .reset_index(drop=True)
)

print('Full results table:')
display(results_table)

Full results table:


,model,loss,best_val_f1,test_acc,test_f1
0,inception,ce,0.351039,0.331958,0.344561
1,mobilenet,ce,0.314306,0.295642,0.308090
2,vgg16,ce,0.315458,0.292341,0.302328
3,vgg19,ce,0.304346,0.276659,0.291320
4,inception,smooth_ordinal,0.358432,0.385606,0.362084
5,mobilenet,smooth_ordinal,0.347630,0.379663,0.346552
6,vgg16,smooth_ordinal,0.361775,0.374711,0.335268
7,vgg19,smooth_ordinal,0.345060,0.380984,0.325560


In [18]:
print('Pivot: Test Accuracy and Test F1 by model and loss')
pivot = results_table.pivot(index='model', columns='loss', values=['test_acc', 'test_f1'])
display(pivot)

Pivot: Test Accuracy and Test F1 by model and loss


test_acc                  test_f1               
loss             ce smooth_ordinal        ce smooth_ordinal
model                                                      
inception  0.331958       0.385606  0.344561       0.362084
mobilenet  0.295642       0.379663  0.308090       0.346552
vgg16      0.292341       0.374711  0.302328       0.335268
vgg19      0.276659       0.380984  0.291320       0.325560

In [19]:
print('Best model per loss function:')
for loss_name in results_table['loss'].unique():
    subset = results_table[results_table['loss'] == loss_name]
    best   = subset.loc[subset['test_f1'].idxmax()]
    print(f'  {loss_name:20s} -> {best["model"]:12s} | Test F1: {best["test_f1"]:.4f} | Test Acc: {best["test_acc"]:.4f}')

Best model per loss function:
  ce                   -> inception    | Test F1: 0.3446 | Test Acc: 0.3320
  smooth_ordinal       -> inception    | Test F1: 0.3621 | Test Acc: 0.3856


# 1. Methodology
Embryo stages follow a natural sequence, the labels were treated as **ordinal categories** rather than independent classes. Additionally, class 15 was merged into class 14 due to its very low sample count and strong visual similarity, resulting in a total of 15 effective classes. To prevent data leakage, the dataset was split **embryo-wise**, ensuring that frames from the same embryo do not appear across training, validation, and test sets.

Four pretrained CNN architectures were evaluated:

- MobileNetV2  
- VGG16  
- VGG19  
- InceptionV3  

For each model, two training settings were considered:

- **Baseline:** Weighted CrossEntropy loss with label smoothing  
- **Custom loss:** Smooth Ordinal Loss  

Predicting a nearby stage is less severe than predicting a distant stage. Standard CrossEntropy does not capture this structure, as it only focuses on maximizing the probability of the correct class.

To address this, the Smooth Ordinal Loss extends CrossEntropy by adding a penalty based on the **distance between predicted and true stage indices**. The key idea is to compute a soft prediction:

$$
y_{\text{soft}} = \sum_{k=0}^{C-1} k \cdot p_k
$$

where $p_k$ is the softmax probability for class $k$. This provides a differentiable estimate of the predicted stage.

The final loss is defined as:

$$
L = L_{CE} + \alpha \left( e^{\frac{|y_{\text{soft}} - y|}{\tau}} - 1 \right)
$$

where:

- $L_{CE}$ is the weighted CrossEntropy loss  
- $|y_{\text{soft}} - y|$ measures ordinal distance  
- $\alpha$ controls the strength of the penalty  
- $\tau$ controls how quickly the penalty grows  

The loss is fully differentiable since it is composed of softmax, summation, and exponential functions, allowing it to be optimized using standard backpropagation.

---

# 2. Results and Analysis
Both CrossEntropy and Smooth Ordinal Loss were evaluated across all four models using test accuracy and weighted F1-score.

The results show a clear and consistent trend: **Smooth Ordinal Loss improves performance across all architectures**. In every model, both test accuracy and F1-score increased when compared to the standard CrossEntropy baseline.

The best overall performance was achieved by **InceptionV3 with Smooth Ordinal Loss**, which obtained the highest test accuracy (0.3856) and test F1-score (0.3621). This indicates that combining a strong architecture with an ordinal-aware loss function leads to better learning of embryo stage progression.

Significant improvements were also observed for the other models. For example, **MobileNetV2** showed a notable increase in both accuracy and F1-score when using Smooth Ordinal Loss, despite being a lightweight architecture. Similarly, **VGG16 and VGG19** both benefited from the ordinal loss, with consistent gains in performance over the CrossEntropy baseline. By penalizing predictions based on their distance from the true stage, the model learns more meaningful representations.

---

# 3. Conclusion

This work demonstrates that treating embryo stage classification as an **ordinal problem** leads to improved performance compared to standard classification approaches.

The proposed Smooth Ordinal Loss effectively incorporates stage-order information by penalizing predictions based on their distance from the true stage. This results in more meaningful learning, particularly for models capable of capturing complex visual patterns.

Among all configurations, **InceptionV3 combined with Smooth Ordinal Loss achieved the best performance**, making it the most suitable approach for this task.

In conclusion, the study highlights that aligning the loss function with the inherent structure of the problem is crucial. Incorporating ordinal relationships into deep learning models can significantly improve performance in tasks where class labels follow a natural progression, such as embryo development stages.